hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/clean/CleanPlanActionExecutor.java

```diff
@@ -97,6 +99,64 @@ private boolean needsCleaning(CleaningTriggerStrategy strategy) {
     }
   }
 
+  private HoodieCleanerPlan getEmptyCleanerPlan(Option<HoodieInstant> earliestInstant, CleanPlanner<T, I, K, O> planner) throws IOException {
+    LOG.info("Nothing to clean here. It is already clean");
+    Option<HoodieInstant> instantVal = getEarliestCommitToRetainToCreateEmptyCleanPlan(earliestInstant);
+    HoodieCleanerPlan.Builder cleanBuilder = HoodieCleanerPlan.newBuilder();
+    instantVal.map(x -> cleanBuilder.setPolicy(config.getCleanerPolicy().name())
+        .setVersion(CleanPlanner.LATEST_CLEAN_PLAN_VERSION)
+        .setEarliestInstantToRetain(new HoodieActionInstant(x.getTimestamp(), x.getAction(), x.getState().name()))
+        .setLastCompletedCommitTimestamp(planner.getLastCompletedCommitTimestamp())
+    ).orElse(cleanBuilder.setPolicy(HoodieCleaningPolicy.KEEP_LATEST_COMMITS.name()));
+    return cleanBuilder.build();
+  }
+
+  /**
+   * There is nothing to clean so this method will create an empty clean plan. But there is chance of optimizing
+   * the subsequent cleaner calls. Consider this scenarios in incremental cleaner mode,
+   * If clean timeline is empty or no clean commits were created for a while then every clean call will have to
+   * scan all the partitions, by creating an empty clean commit to update earliestCommitToRetain instant value,
+   * incremental clean policy does not have to look for file changes in all the partitions, rather it will look
+   * for partitions that are modified in last x hours. This value is configured through MAX_DURATION_TO_CREATE_EMPTY_CLEAN.
```

Comment:
-1, I don't think it deserves to introduce sucn complexity. Basically we need to support efficient partitions query based on the metadata(either from timeline or MDT), that I thought is the way to go.

Reply:
@danny0405 
The usecase we are trying to solve here is when a dataset that does not receive upserts and does not use small handling, cleans are running for every commit and scans all the partitions all the time. Also, we cannot disable cleans for such datasets, since those datasets can be onboarded to clustering and can do adhoc rewrites on some partitions. So, clean have to be enabled all the time and when we are enabling cleans it is scanning all the partitions. So, we thought of storing the earliestCommitToRetain value somewhere and use it to store the commits it has scanned. That way every clean will only scan the partitions that were touched in a commit since earliestCommitToRetain progress commit by commit when incremental clean policy is enabled. This reduces the full scan for all partitions to scanning one or two partitions.
CC @nsivabalan

Reply:
> Also, we cannot disable cleans for such datasets, since those datasets can be onboarded to clustering and can do adhoc rewrites on some partitions. So, clean have to be enabled all the time and when we are enabling cleans it is scanning all the partitions. So, we thought of storing the earliestCommitToRetain value somewhere and use it to store the commits it has scanned. 

Can we do force clean in the rewrite jobs or store the last `earliestCommitToRetain` separately in the metadata, instead of empty clean commits, like as you mentioned, most of the clean commits would just be useless and empty for this special inserts scenario. I kind of think we should come up with a more general solution for append only use case.

Reply:
> Can we do force clean in the rewrite jobs
I think this may not work since we need to have restore capabilities on the dataset for 24 hours.
> store the last earliestCommitToRetain separately in the metadata
Generally speaking, clean's earliestCommitToRetain value and dataset's checkpoints information should be stored in a separate location maybe hoodie.properties or something like in a Hudi metaserver. Do you think we should create separate partition in metadata table for storing these such key, value configs?

Reply:
> Do you think we should create separate partition in metadata table for storing these such key, value configs?

I kind of think storing the single value `earliestCommitToRetain`somewhere in the .hoodie is feasible, just not sure whether we should utilize the MDT because MDT itself is very heavy-weight, maybe some auxiliary marker file is enough.

Reply:
hey Danny. I am trying to understand your suggestion 
```
Can we do force clean in the rewrite jobs or store the last earliestCommitToRetain separately in the metadata, instead of empty clean commits,
```

not sure how we can enforce force clean. Are you suggesting after clean planning is complete, even if there is nothing to clean, lets add a clean commit to timeline. 

I don't think we can go that route. then it will double the timeline for workloads which might have just updates w/ MOR table where cleaner may not have something to clean everytime.

Reply:
I also do not like the idea of creating empty clean commits. But can't think of any other elegant way. 
here in this patch, what surya is trying to do is, enforce empty clean if there are no clean commits in timeline for extended period of time( say 12 hours). So every 12 hours we might end up creating an empty clean commit to timeline which is not a bad idea. 

happy to jam more to find some elegant way.

Reply:
@nsivabalan  more like one commit every 24 hours, it is configurable though.
Another approach is to store move the earliestCommitToRetain and deltstreamer.checkpoints from the commit files and store it in a separate file other than the hoodie.properties file. Since, rewriting hoodie.properties multiple times will also cause some issues while querying. I can setup a call early next week to jam on this issue.

Reply:
hive meta sync also use `last_commit_completion_time_sync` to reduce scan all partition path, it store in hms.

Reply:
hive meta sync also use last_commit_completion_time_sync to reduce scan all partition path, it store in hms. Can be used as a reference.